File: src/main/java/de/fluxparticle/animation/graph/GraphWrapper.java
package de.fluxparticle.animation.graph;

import de.fluxparticle.utils.chain.Chain;

import java.util.*;

import static de.fluxparticle.utils.chain.Chain.cons;
import static de.fluxparticle.utils.chain.Chain.emptyChain;

public class GraphWrapper implements Graph {

	private final Map<Integer, GraphNode> nodes = new HashMap<>();

	private final Map<Integer, Set<Integer>> edges = new HashMap<>();

	public GraphWrapper(Graph graph) {
        Collection<? extends GraphNode> nodes = graph.getNodes();
        for (GraphNode node : nodes) {
            depthFirst(graph, node, emptyChain());
        }
    }

    private void depthFirst(Graph graph, GraphNode node, Chain<GraphNode> chain) {
        if (nodes.containsKey(node.getId())) {
            return;
        }

        Integer id = node.getId();
        nodes.put(id, node);

        HashSet<Integer> edgeSet = new HashSet<>();
        for (GraphNode next : graph.getNeighbours(node)) {
            if (chain.asStream().noneMatch(next::equals)) {
                edgeSet.add(next.getId());
                depthFirst(graph, next, cons(next, chain));
            }
        }

        edges.put(id, edgeSet);
    }

    @Override
	public Collection<GraphNode> getNodes() {
		return Collections.unmodifiableCollection(nodes.values());
	}

	@Override
	public Collection<GraphNode> getNeighbours(GraphNode node) {
		int nodeId = node.getId();

		Collection<GraphNode> neighbours = new LinkedList<>();

		for (Integer neighbourId : edges.get(nodeId)) {
			neighbours.add(nodes.get(neighbourId));
		}

		return neighbours;
	}

}

File: src/main/kotlin/de/fluxparticle/animation/graph/GraphWrapper.kt
package de.fluxparticle.animation.graph

import de.fluxparticle.utils.chain.Chain
import de.fluxparticle.utils.chain.chainOf
import de.fluxparticle.utils.chain.cons

class GraphWrapper(graph: Graph) : Graph {

    private val nodeMap = HashMap<Int, GraphNode>()

    private val edges = HashMap<Int, Set<Int>>()

    override val nodes: Collection<GraphNode>
        get() = nodeMap.values

    init {
        graph.nodes.forEach { node ->
            depthFirst(graph, node, chainOf())
        }
    }

    private fun depthFirst(graph: Graph, node: GraphNode, chain: Chain<GraphNode>) {
        if (nodeMap.containsKey(node.id)) {
            return
        }

        val id = node.id
        nodeMap.put(id, node)

        val edgeSet = HashSet<Int>()
        for (next in graph.getNeighbours(node)) {
            if (chain.none { next == it }) {
                edgeSet.add(next.id)
                depthFirst(graph, next, cons(next, chain))
            }
        }

        edges.put(id, edgeSet)
    }

    override fun getNeighbours(node: GraphNode): Collection<GraphNode> {
        return edges[node.id]!!.map { neighbourId -> nodeMap[neighbourId]!! }
    }

}
